src/java/apiview-java-processor/src/main/java/com/azure/tools/apiview/processor/diagnostics/Diagnostics.java

```diff
@@ -40,25 +47,36 @@ public Diagnostics() {
         diagnostics.add(new FluentSetterReturnTypeDiagnosticRule());
         diagnostics.add(new ConsiderFinalClassDiagnosticRule());
         diagnostics.add(new IllegalMethodNamesDiagnosticRule(
-            new Rule("Builder$", "tokenCredential"), // it should just be 'credential'
-            new Rule("Builder$", "^set"),            // we shouldn't have setters in the builder
-            new Rule("^isHas"),
-            new Rule("^setHas")
+                new Rule("Builder$", "tokenCredential"), // it should just be 'credential'
+                new Rule("Builder$", "^set"),            // we shouldn't have setters in the builder
+                new Rule("^isHas"),
+                new Rule("^setHas")
         ));
         diagnostics.add(new MissingJavaDocDiagnosticRule());
         diagnostics.add(new MissingJavadocCodeSnippetsRule());
         diagnostics.add(new NoLocalesInJavadocUrlDiagnosticRule());
         diagnostics.add(new ModuleInfoDiagnosticRule());
+        diagnostics.add(new ServiceVersionDiagnosticRule());
 
         // common APIs for all builders (below we will do rules for http or amqp builders)
         diagnostics.add(new RequiredBuilderMethodsDiagnosticRule(null)
             .add("configuration", new ExactTypeNameCheckFunction("Configuration"))
             .add("clientOptions", new ExactTypeNameCheckFunction("ClientOptions"))
             .add("connectionString", new ExactTypeNameCheckFunction("String"))
             .add("credential", new ExactTypeNameCheckFunction(new ParameterAllowedTypes("TokenCredential",
-                    "AzureKeyCredential", "AzureSasCredential")))
+                    "AzureKeyCredential", "AzureSasCredential", "AzureNamedKeyCredential")))
             .add("endpoint", new ExactTypeNameCheckFunction("String"))
-            .add("serviceVersion", new DirectSubclassCheckFunction("ServiceVersion")));
+            .add("serviceVersion", methodDeclaration -> {
+                Type parameterType = methodDeclaration.getParameter(0).getType();
+                ClassOrInterfaceType classOrInterfaceType = parameterType.asClassOrInterfaceType();
+                if (!classOrInterfaceType.getNameAsString().endsWith("ServiceVersion")) {
+                    return Optional.of(
+                            new Diagnostic(WARNING, makeId(methodDeclaration),
+                                    "Incorrect type being supplied to this builder method. Expected an enum "
+                                    + "implementing ServiceVersion but was " + classOrInterfaceType.getNameAsString() + "."));
+                }
+                return Optional.empty();
+            }));
```

Comment:
I'd like to keep this section shorter and succinct - we should move this out to a separate function that we method reference.